<launch>

<!-- initial pose -->
<arg name="initialpose_x" value="0"/>
<arg name="initialpose_y" value="0"/>
<arg name="initialpose_a" value="0"/>
<arg name="initial_state" value="[$(arg initialpose_x), $(arg initialpose_y), 0, 0, 0, $(arg initialpose_a), 0, 0, 0, 0, 0, 0, 0, 0, 0]"/>

<!-- frame name -->
<arg name="global_frame" value="map1"/>
<arg name="odom_frame"   value="odom1"/>
<arg name="base_frame"   value="base_footprint1"/>
<arg name="scan_frame"   value="base_scan1"/>

<group ns="portable1">

  <!-- map server -->
  <arg name="map_file" value="$(find tms_rc_bot)/maps/huistenbosch_windmill/map.yaml"/>
  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)">
    <param name="frame_id" value="$(arg global_frame)"/>
  </node>

  <!-- static tf -->
  <node pkg="tf" type="static_transform_publisher" name="world_to_map"
    args="0 0 0 0 0 0 world_link $(arg global_frame) 100">
  </node>
  
  <node pkg="tf" type="static_transform_publisher" name="map_to_measured"
    args="0 0 0 0 0 0 $(arg global_frame) base_measured 100">
  </node>

  <node pkg="tf" type="static_transform_publisher" name="foot_to_scan"
    args="0.45 -0.22 0.1 0 0 0 $(arg base_frame) $(arg scan_frame) 100">
  </node>
  
  <!-- move base -->
  <arg name="move_base" default="true"/>
  <node pkg="move_base" type="move_base" name="move_base" respawn="false" output="screen" if="$(arg move_base)">
    <remap from="odom"                            to="/portable1/odom"/>
    <param name="global_costmap/global_frame"     value="$(arg global_frame)"/>
    <param name="global_costmap/robot_base_frame" value="$(arg base_frame)"/>
    <param name="local_costmap/global_frame"      value="$(arg global_frame)"/>
    <param name="local_costmap/robot_base_frame"  value="$(arg base_frame)"/>
    <rosparam file="$(find tms_rc_bot)/param/chairbot/move_base.yaml"/>
    <rosparam file="$(find tms_rc_bot)/param/chairbot/global_planner.yaml"/>
    <rosparam file="$(find tms_rc_bot)/param/chairbot/local_planner.yaml"/>
    <rosparam file="$(find tms_rc_bot)/param/chairbot/costmap_separate.yaml"/>
    <rosparam file="$(find tms_rc_bot)/param/chairbot/costmap_common.yaml" ns="global_costmap"/>
    <rosparam file="$(find tms_rc_bot)/param/chairbot/costmap_common.yaml" ns="local_costmap"/>
  </node>

  <!-- cmd vel mux -->
  <node pkg="rostopic" type="rostopic" name="safe" args="pub safety_twist geometry_msgs/Twist '[0, 0, 0]' '[0, 0, 0]' -r 10"/>
  <node args="standalone yocs_cmd_vel_mux/CmdVelMuxNodelet" pkg="nodelet" type="nodelet" name="cmd_vel_mux">
    <param name="yaml_cfg_file"  value="$(find tms_rc_bot)/param/common/cmd_vel_mux.yaml"/>
    <remap from="~input/keyop"   to="keyop"/>
    <remap from="~input/remote"  to="remote"/>
    <remap from="~input/cmd_vel" to="cmd_vel_smoother"/>
    <remap from="~input/safety"  to="safety_twist"/>
    <remap from="~output"        to="socket_twist"/>
  </node>

  <!-- urg node -->
  <arg name="urg_range" value="1.5708"/>
  <node pkg="urg_node" type="urg_node" name="urg_node" output="screen">
    <remap from="scan"       to="urg_scan"/>
    <param name="ip_address" value="192.168.0.10"/>
    <param name="frame_id"   value="$(arg scan_frame)"/>
    <param name="angle_min"  value="-$(arg urg_range)"/>
    <param name="angle_max"  value="$(arg urg_range)"/>
  </node>

  <!-- portable laser -->
  <node pkg="tms_rc_bot" type="portable_laser" name="portable_laser" output="screen">
    <param name="input_scan_name"  value="urg_scan"/>
    <param name="output_scan_name" value="scan"/>
    <param name="range_min"        value="0.1"/>
    <param name="range_max"        value="20"/>
  </node>

  <!-- portable indicator -->
  <node pkg="tms_rc_bot" type="portable_indicator" name="portable_indicator" output="screen" if="$(arg move_base)"/>

  <!-- portable position -->
  <node pkg="tms_rc_bot" type="portable_position" name="portable_position" output="screen">
    <param name="odom_topic_name" value="/portable1/odom"/>
    <param name="global_frame_id" value="$(arg global_frame)"/>
    <param name="base_frame_id"   value="$(arg base_frame)"/>
    <param name="marker_name"     value="chairbot"/>
    <param name="print_marker"    value="true"/>
  </node>

  <!-- portable measured -->
  <arg name="calibration" default="true"/>
  <node pkg="tms_rc_bot" type="portable_measured" name="portable_measured" output="screen">
    <rosparam file="$(find tms_rc_bot)/param/chairbot/c94_m8p_origin.yaml"/>
    <param name="gps_subscribe"           value="true"/>
    <param name="float_subscribe"         value="false"/>
    <param name="measured_out_topic_name" value="odometry/measured"/>
    <param name="people_topic_name"       value="/people_ninebot_cropped"/>
    <param name="global_frame_id"         value="$(arg global_frame)"/>
    <param name="people_frame_id"         value="$(arg global_frame)"/>
    <param name="print_tf"                value="false" if="$(arg calibration)"/>
  </node>

  <!-- portable wiiremote -->
  <node pkg="tms_rc_bot" type="portable_wiiremote" name="portable_wiiremote" output="screen">
    <param name="cmd_vel_topic" value="wii_cmd"/>
    <param name="linear_active" value="true"/>
    <param name="max_trans_vel" value="0.4"/>
    <param name="max_rot_vel"   value="0.8"/>
  </node>

  <!-- portable smoother remote -->
  <node pkg="tms_rc_bot" type="portable_smoother" name="portable_smoother_remote" output="screen">
    <param name="input_topic_name"  value="wii_cmd"/>
    <param name="output_topic_name" value="remote"/>
    <param name="acc_lim_xy"        value="3.25"/>
    <param name="acc_lim_th"        value="1.25"/>
    <param name="time_out"          value="0.5"/>
    <param name="ignore_stop"       value="false"/>
  </node>

  <!-- portable smoother cmd-->
  <node pkg="tms_rc_bot" type="portable_smoother" name="portable_smoother_cmd" output="screen">
    <param name="input_topic_name"  value="cmd_vel"/>
    <param name="output_topic_name" value="cmd_vel_smoother"/>
    <param name="acc_lim_xy"        value="3.25"/>
    <param name="acc_lim_th"        value="1.25"/>
    <param name="time_out"          value="0.5"/>
    <param name="ignore_stop"       value="false"/>
  </node>

  <!-- chairbot driver -->
  <node pkg="tms_rc_bot" type="chairbot_driver" name="chairbot_driver" output="screen">
    <remap from="odom"          to="odometry/wheel"/>
    <remap from="cmd_vel"       to="socket_twist"/>
    <param name="ip_address"    value="192.168.11.99"/>
    <param name="spin_Kp"       value="1000"/>
    <param name="spin_Ki"       value="100"/>
    <param name="spin_Kd"       value="0"/>
    <param name="odom_frame_id" value="$(arg odom_frame)"/>
    <param name="base_frame_id" value="$(arg base_frame)"/>
    <param name="wheel_radius"  value="0.184"/>
    <param name="wheel_tread"   value="0.544"/>
    <param name="max_trans_vel" value="0.4"/>
    <param name="max_rot_vel"   value="0.8"/>
    <param name="print_tf"      value="true"/>  
  </node>

  <!-- chairbot speaker -->
  <node pkg="tms_rc_bot" type="chairbot_speaker" name="chairbot_speaker" output="screen"/>

  <!-- chairbot calibration -->
  <node pkg="tms_rc_bot" type="chairbot_calibration" name="chairbot_calibration" output="screen" if="$(arg calibration)">
    <rosparam file="$(find tms_rc_bot)/param/chairbot/c94_m8p_origin.yaml"/>
    <param name="global_frame_id" value="$(arg global_frame)"/>
    <param name="file_name"       value="c94_m8p_origin.yaml"/>
  </node>

  <!-- ublox_gps -->
  <node pkg="ublox_gps" type="ublox_gps" name="ublox_gps" output="screen">
    <rosparam file="$(find tms_rc_bot)/param/chairbot/c94_m8p_rover.yaml"/>
    <param name="device" value="/dev/ttyACM1"/>
    <!-- <param name="frame_id" value="base_gps"/> -->
  </node>

  <!-- map_to_odom -->
  <rosparam file="$(find tms_rc_bot)/param/chairbot/robot_local_separate.yaml"/>
  <node pkg="robot_localization" type="ekf_localization_node" name="map_to_odom" output="screen">
    <rosparam file="$(find tms_rc_bot)/param/chairbot/robot_local_common.yaml"/>
    <rosparam param="initial_state" subst_value="True">$(arg initial_state)</rosparam>
    <remap from="set_pose"          to="initialpose"/>
    <param name="map_frame"         value="$(arg global_frame)"/>
    <param name="odom_frame"        value="$(arg odom_frame)"/>    
    <param name="base_link_frame"   value="$(arg base_frame)"/>
    <param name="world_frame"       value="$(arg global_frame)"/>
  </node>

  <!-- odom_to_base -->
  <!-- <node pkg="robot_localization" type="ekf_localization_node" name="odom_to_base" output="screen">
    <rosparam file="$(find tms_rc_bot)/param/chairbot/robot_local_common.yaml"/>
    <rosparam param="initial_state" subst_value="True">$(arg initial_state)</rosparam>
    <remap from="set_pose"          to="initialpose"/>
    <param name="map_frame"         value="$(arg global_frame)"/>
    <param name="odom_frame"        value="$(arg odom_frame)"/>    
    <param name="base_link_frame"   value="$(arg base_frame)"/>
    <param name="world_frame"       value="$(arg odom_frame)"/>
  </node> -->

  <!-- fake_localization -->
  <!-- <node pkg="fake_localization" type="fake_localization" name="fake_localization" output="screen">
    <remap from="base_pose_ground_truth" to="odometry/wheel"/>
    <param name="global_frame_id"        value="$(arg global_frame)"/>
    <param name="odom_frame_id"          value="$(arg odom_frame)"/>
    <param name="base_frame_id"          value="$(arg base_frame)"/>
    <param name="initial_pose_x"         value="$(arg initialpose_x)"/>
    <param name="initial_pose_y"         value="$(arg initialpose_y)"/>
    <param name="initial_pose_a"         value="$(arg initialpose_a)"/>
  </node> -->

  <!-- amcl -->
  <!-- <node pkg="amcl" type="amcl" name="amcl" output="screen">
    <rosparam file="$(find tms_rc_bot)/param/chairbot/robot_local_amcl.yaml"/>
    <param name="global_frame_id" value="$(arg global_frame)"/>
    <param name="odom_frame_id"   value="$(arg odom_frame)"/>
    <param name="base_frame_id"   value="$(arg base_frame)"/>
    <param name="initial_pose_x"  value="$(arg initialpose_x)"/>
    <param name="initial_pose_y"  value="$(arg initialpose_y)"/>
    <param name="initial_pose_a"  value="$(arg initialpose_a)"/>
  </node> -->

</group>
</launch>
